### PR TITLE
Update quality doc with guidelinse for good unit tests

### PR DESCRIPTION
**Pull request by mmislam101**

### CODE REVIEW

**Comment by swilliams** [comment]
@mmislam101 I jut saw this, will dive in. Thank you.

File: docs/Quality.md [hunk 1]
```diff
@@ -11,18 +11,45 @@ There is a delicate balance between doing great work and actually shipping somet
...
-This is our first line of defense. While it is possible to go overboard with testing zealotry, there is great value in _automated_ tests that run against your code. Tests fall into different categories. 
+This is our first line of defense. While it is possible to go overboard with testing zealotry, there is great value in _automated_ tests that run against your code. Tests fall into different categories.
 
 ### Unit Tests
 A unit test examines a single function or method (a single "unit"). For example, the tests for a function called `addIntegers` would make sure that the results of that function are the sum of the arguments passed in. These are the easiest to write, but also provide the least amount of utility because units of code never run in isolation. More bugs occur when the units meet, or to use the proper nomenclature, when they integrate with one another.
 
+#### Criteria for useful unit tests
+1. Will it *really* catch any bugs?
+2. Will it prevent developers from injecting new bugs?
+3. Will it strengthen un-avoidable brittle code?
```
Comment (swilliams) [inline docs/Quality.md]: 'unavoidable' isn't hyphenated.

File: docs/Quality.md [hunk 2]
```diff
@@ -11,18 +11,45 @@ There is a delicate balance between doing great work and actually shipping somet
...
+4. Will it help maintain un-avoidable complex algorithms?
+5. Will it test real corner cases?
+  - A real corner case is when there's a greater-than-zero chance of something occurring in the field
+- An un-real corner case is execution of the default of a switch statement that will never occur unless a solar flare flips a bit on a physical register
+6. For weakly typed languages, will it help maintain correct type usage?
+
+The following are helpful, but have short expiration dates, and quickly becomes a hindrance, or toxic, to development.
```
Comment (swilliams) [inline docs/Quality.md]: 'becomes' -> 'become', and I'd nix the 'or toxic' part.

File: docs/Quality.md [hunk 3]
```diff
@@ -11,18 +11,45 @@ There is a delicate balance between doing great work and actually shipping somet
...
+1. Will it help catch errors when you refactor?
+2. Will it help you during development (TDD)? Will it get from point A to Z? Will it help highlight any holes in your logic, and force you to think of corner cases?
```
Comment (swilliams) [inline docs/Quality.md]: Are arguing that tests that catch errors during refactoring should be removed?

File: docs/Quality.md [hunk 4]
```diff
@@ -11,18 +11,45 @@ There is a delicate balance between doing great work and actually shipping somet
...
+It can be argued that these tests should be removed, or disabled, so that it doesn't slow down development.
+
+#### Indicators of useless, or toxic, unit tests
+1. Is it just testing the language. For example, is it testing whether a switch statement is working?
```
Comment (swilliams) [inline docs/Quality.md]: Remove 'just'

File: docs/Quality.md [hunk 5]
```diff
@@ -11,18 +11,45 @@ There is a delicate balance between doing great work and actually shipping somet
...
+2. Is it testing an already vetted, well used, library or function? Example, testing the sqrt() function in the math.lib
+3. Does the cost of maintaining the unit test far outweigh the benefits of keeping the test around?
+4. Is the test brittle and prone to show false-negatives?
+
+#### Examples of good unit tests
+1. Test to maintain complex algorithms that calculates the recommendation of contacts that could be your family. The tests makes sure that any changes to the algorithm, that affects expected recommendation results, causes tests to fail. The tests were also helpful as TDD, to get from a list of contacts (point A) to the expected recommendations (point Z). The tests were also helpful post-production, where each corner case found on the field had a new test to future-proof against regression.
```
Comment (swilliams) [inline docs/Quality.md]: Pluralization gets a little weird here. "Tests makes" etc.

**Comment by mmislam101** [comment]
@swilliams Added more details around my controversial opinions.
